### DISCUSSION TITLE
How to use MoshiConverterFactory for generic type of model classes in android ?

### DISCUSSION THREAD

**Discussion by GopalLR** [square/moshi#1721] [General]
Hi, Can I use MoshiConverterFactory and GsonConverterFactory in single retrofit instance ?

I'm working in one project which has morethen 50 APIs and all APIs response structure is same like below :

{

  "status": 200,

  "message": "Login successfully.",

  "result": {

  }

}

I create base model class with generic type

@JsonClass(generateAdapter = true)

data class BaseModel<T>(

    @Json(name = "status")

    val status: Int,

    @Json(name = "message")

    val message: String,

    @Json(name = "result")

    val result: Map<String, T>

)

@JsonClass(generateAdapter = true)

data class AccessToken(

    @Json(name = "accessToken")

    val accessToken: String

)

It's give error like : 

Unable to create converter for com.app.sample.data.model.BaseModel<com.app.sample.data.model.AccessToken> for method AuthService.loginUser in android

How to fixed this issue using MoshiConverterFactory ?

I look forward to hearing from you.

**Comment by ZacSweers** [comment]
This is not the Retrofit issue tracker. Please ask questions in its discussions section here: https://github.com/square/retrofit/discussions

**Reply by GopalLR** [reply]
Already added in discussion section 

https://github.com/square/retrofit/discussions?discussions_q=1721

**Comment by JakeWharton** [comment]
This is not a problem with Retrofit. Moshi is failing to create an adapter for the supplied type.

However, you haven't provided the full stacktrace of the failure so it's impossible to know what's going on. Going to move to a discussion here.

**Comment by C2418M31** [comment]
I'm trying to use @Json(name = "Result") property to change the field name and it's not working. The result is always null from retrofit response, but if I use the actual "Result" in the field it is working fine, here's the example data class.

What am I missing?

**There is no crash it just result to null**

Not Working:
`@JsonClass(generateAdapter = true)`
`data class SMSResponse(`
`    @Json(name = "Result")`
`    val smsName: String?`
`)`

Also Not working when using @field:Json(name = "Result")
`@JsonClass(generateAdapter = true)`
`data class SMSResponse(`
`    @field:Json(name = "Result")`
`     val smsName: String?`
`)`

Working:
`@JsonClass(generateAdapter = true)`
`data class SMSResponse(`
`   val Result: String?`
`)`

NetworkCall:
`return Retrofit.Builder()`
`        .addConverterFactory(GsonConverterFactory.create(gson))`
`        .addConverterFactory(MoshiConverterFactory.create())`
`        .baseUrl(BuildConfig.BASE_URL)`
`        .client(okHttpClient)`
`        .build()`

Dependency:
`implementation("com.squareup.moshi:moshi-kotlin:1.14.0")`
`ksp("com.squareup.moshi:moshi-kotlin-codegen:1.15.0")`


UPDATE:

Got it fixed by removing the ` .addConverterFactory(GsonConverterFactory.create(gson))`. It seems you cannot have two Converter
